### PR TITLE
Feat(api-platform): handle scope on endpoints

### LINKED ISSUE #33242: Handle scopes on endpoints
### Docs
Doc Symfony cast roles https://symfonycasts.com/screencast/api-platform-security/scopes
API Platform security https://api-platform.com/docs/core/security/
https://symfonycasts.com/screencast/api-platform-security/access-control-voter

### Default security configurations

The default behaviour of APIPlatform already provides some ways to configure our scopes both for security reasons and to integrate them in our API doc but they may be a bit hard to use:

```php
    new Get(
            uriTemplate: '/hook-status/{hookId}',
            requirements: ['id' => '\d+'],
            openapiContext: [
                'summary' => 'Get hook status A',
                'description' => 'Get hook status B',
                'parameters' => [
                    [
                        'name' => 'hookId',
                        'in' => 'path',
                        'required' => true,
                        'schema' => [
                            'type' => 'string',
                        ],
                        'description' => 'Id of the hook you are requesting the status from',
                    ],
                    [
                        'name' => 'Authorization',
                        'in' => 'scopes',
                        'description' => 'read:hook-status <br> write:hook-status ',
                    ],
                ],
            ],
            security: 'is_granted("read:hook-status") OR is_granted("write:hook-status")'
            exceptionToStatus: [HookNotFoundException::class => 404],
            provider: QueryProvider::class,
            extraProperties: ['query' => GetHookStatus::class]
        ),
```

Security:
`security: 'is_granted("read:hook-status") OR is_granted("write:hook-status")'`

Doc:
```php
                    [
                        'name' => 'Authorization',
                        'in' => 'scopes',
                        'description' => 'read:hook-status <br> write:hook-status ',
                    ],
```

### Wanted
We would like to make this configuration a bit simpler and use an array of scopes so the configuration would look more like this:

```php
    new Get(
            uriTemplate: '/hook-status/{hookId}',
            requirements: ['id' => '\d+'],
            openapiContext: [
                'summary' => 'Get hook status A',
                'description' => 'Get hook status B',
                'parameters' => [
                    [
                        'name' => 'hookId',
                        'in' => 'path',
                        'required' => true,
                        'schema' => [
                            'type' => 'string',
                        ],
                        'description' => 'Id of the hook you are requesting the status from',
                    ],
                ],
            ],
            scopes: ['read:hook-status', 'write:hook-status'],
            exceptionToStatus: [HookNotFoundException::class => 404],
            provider: QueryProvider::class,
            extraProperties: ['query' => GetHookStatus::class]
        ),
```

The scopes would be defined by this single line: `scopes: ['read:hook-status', 'write:hook-status']`

It can be seen as a shortcut a syntaxic sugar, then it would be our job to correctly interpret this in two ways:
- this setting must be translated into security rules and have the same behaviour they would have `security: 'is_granted("read:hook-status") OR is_granted("write:hook-status")'` We need to search how this would be done, is it intepeted by a ContainerPass? is there a listener in the APIPlatform that allows to add some rules? how is the `security` field of APIPlatform handled right now, and can we integrate ourselves in it? maybe there's a solution to add a new custom type of Voter (see doc at the beginning) These are a few leads that need to be looked into to see if we can reach this goal
- if the first has found a solution (and only then), we could look into the APIDoc system to see if those scopes can be added automatically in an endpoint description This would prevent having to manually set it in the `openapiContext` field

**Bonus:**

Absolutely not required but it could be interesting to handle some kind of hierarchy:

```yaml
- order_read
  - order_invoice_read
  - order_products_read
- order_write
   - order_invoice_write
   - order_products_write
- order_read_write
  - over_read
  - order_write
```

It seems like Symfony already handles this: https://symfony.com/doc/current/security.html#hierarchical-roles But to be tested, and see how this would influence the two previous points (it should work out of the bo regarding security rules, but it may be a bit more challenging for the API doc)

### UI Tests

Please contact the @PrestaShop/ui-tests-maintainers team when the PR is available and before sending it to @PrestaShop/qa-functional for testing.

### PR DESCRIPTION
**Pull request by tleon**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | <ul><li>Modified the auth server to return scopes</li><li> Added thoses scopes to the in memory user</li><li>Configured api platform to integrate those scopes</li><li>Decorated an api platform service to transform extra property "scopes" into a security rule</li></ul>
| Type?             | improvement 
| Category?         | BO 
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | <ul><li>Get an bearer token by requesting the /api/oauth2/token endpoint</li><li> Request the /api/hook-status/{id} or the /api/hooks/{id} endpoints </li><li>All green https://github.com/tleon/ga.tests.ui.pr/actions/runs/6224787955/job/16893766740</li></ul>
| Fixed issue or discussion?     | Fixes #33242 
| Sponsor company   | PrestaShop SA

More informations on how to get your bearer token :

You need to add a return statement at the start of the onKernelRequest method of the SslMiddleware.php file. If you don't do this postman won't be able to handle the current ssl protocol.

### CODE REVIEW

File: app/config/security_dev.yml
Comment (mflasquin) [inline app/config/security_dev.yml]: Do not push that ;)

File: src/PrestaShopBundle/Security/OAuth2/Repository/NullScopeRepository.php
Comment (mflasquin) [inline src/PrestaShopBundle/Security/OAuth2/Repository/NullScopeRepository.php]: I don't understand why we mention hook:read here

**Review by mflasquin — DISMISSED** [review]

File: src/PrestaShopBundle/Security/OAuth2/Repository/NullScopeRepository.php
Comment (tleon) [reply src/PrestaShopBundle/Security/OAuth2/Repository/NullScopeRepository.php]: When we looked into this it seemed to be required to be able to have the scopes into the token.

File: app/config/security_dev.yml
Comment (tleon) [reply app/config/security_dev.yml]: I also talked about this with @jolelievre but I still don't really understant why this is a problem since it's only for dev env.

Comment (mflasquin) [reply app/config/security_dev.yml]: https://github.com/PrestaShop/PrestaShop/issues/31694

File: src/PrestaShopBundle/Security/OAuth2/Repository/NullScopeRepository.php
Comment (mflasquin) [reply src/PrestaShopBundle/Security/OAuth2/Repository/NullScopeRepository.php]: maybe we can add scopes in yaml config ?

File: src/PrestaShopBundle/ApiPlatform/Decorator/AttributesResourceMetadataCollectionFactoryDecorator.php
Comment (jolelievre) [inline src/PrestaShopBundle/ApiPlatform/Decorator/AttributesResourceMetadataCollectionFactoryDecorator.php:38]: We should add a description here, this decoration principle is not easy to grasp we must inform future developers what this is all about as they probably won't be able to deduce from the code on first sight

File: src/PrestaShopBundle/ApiPlatform/Decorator/AttributesResourceMetadataCollectionFactoryDecorator.php
Comment (jolelievre) [inline src/PrestaShopBundle/ApiPlatform/Decorator/AttributesResourceMetadataCollectionFactoryDecorator.php]:
```suggestion
                    $modifiedOperationWithSecurity = $operation->withSecurity($this->translateScopeToSecurity($scopesToAdd), $operation->getSecurity());
```

Comment (jolelievre) [reply src/PrestaShopBundle/ApiPlatform/Decorator/AttributesResourceMetadataCollectionFactoryDecorator.php]: I thought maybe the current value should be passed and used/modified in case someone already added a custom alternative rule like `isSuperAdmin() OR is_granted(CUSTOM_ROLE)`

In that idea I thought that if a rule already exists if should be "added" to the rules for scopes with an or condition (see my comment below)

File: src/PrestaShopBundle/ApiPlatform/Decorator/AttributesResourceMetadataCollectionFactoryDecorator.php
Comment (jolelievre) [inline src/PrestaShopBundle/ApiPlatform/Decorator/AttributesResourceMetadataCollectionFactoryDecorator.php]:
```suggestion
    private function translateScopeToSecurity(array $scopes, string $existingSecurity): string
```

File: src/PrestaShopBundle/ApiPlatform/Decorator/AttributesResourceMetadataCollectionFactoryDecorator.php
Comment (jolelievre) [inline src/PrestaShopBundle/ApiPlatform/Decorator/AttributesResourceMetadataCollectionFactoryDecorator.php]:
```suggestion
        return empty($existingSecurity) ? $security : sprintf('(%s) OR (%s)', $existingSecurity, $security);
```

File: src/PrestaShopBundle/ApiPlatform/Resources/Hook.php
Comment (jolelievre) [inline src/PrestaShopBundle/ApiPlatform/Resources/Hook.php]: Ah in the pending ADR https://github.com/PrestaShop/ADR/pull/35 I suggested to use camel case for the domain entity so `Hook:read`

It's not set in stone but I guess it's the good moment to comment and challenge the ADR about this rule so that it's settled and we know our convention

File: src/PrestaShopBundle/Security/OAuth2/Repository/NullScopeRepository.php
Comment (jolelievre) [reply src/PrestaShopBundle/Security/OAuth2/Repository/NullScopeRepository.php]: Shouldn't we create the scope based on the parameter instead?
```suggestion
        return new ScopeEntity($identifier);
```

For now we only return it as is and consider any asked scope exists, but later we'll have to check its existence based on the scopes that were parsed from the API endpoints definitions

File: src/PrestaShopBundle/Security/OAuth2/Repository/NullScopeRepository.php
Comment (jolelievre) [inline src/PrestaShopBundle/Security/OAuth2/Repository/NullScopeRepository.php]:
```suggestion
        return $scopes;
```

Similar suggestion here, we could return the request scopes as is and consider they are all valid for now

File: src/PrestaShopBundle/Security/OAuth2/Repository/ScopeEntity.php
Comment (jolelievre) [inline src/PrestaShopBundle/Security/OAuth2/Repository/ScopeEntity.php]:
```suggestion
namespace PrestaShopBundle\Security\OAuth2\Entity;
```

File: app/config/security_test.yml
Comment (jolelievre) [inline app/config/security_test.yml]: Why is the role written this way capitalized? Is it because there's a system somewhere that automatically uppercases the scope value? (and appends `ROLE_` in front of it?)

I was expecting we could do:

```suggestion
          my_client_id: { password: 'prestashop', roles: [ 'Hook:read' ] }
```

with the name of the scope directly used as the role name. But from what you show here it seems Symfony roles have some mandatory format always in capitalized characters? But which component converts this automatically then?

File: src/PrestaShopBundle/ApiPlatform/Decorator/AttributesResourceMetadataCollectionFactoryDecorator.php
Comment (jolelievre) [inline src/PrestaShopBundle/ApiPlatform/Decorator/AttributesResourceMetadataCollectionFactoryDecorator.php]:
```suggestion
            $security .= 'is_granted("ROLE_' . strtoupper($scope) . '")';
```

File: src/PrestaShopBundle/ApiPlatform/Decorator/AttributesResourceMetadataCollectionFactoryDecorator.php
Comment (jolelievre) [reply src/PrestaShopBundle/ApiPlatform/Decorator/AttributesResourceMetadataCollectionFactoryDecorator.php:38]: Good start for the explanation, but it mostly explains what a decore does "change the result" ^^
Can you please add more detail about what it actually does like `parse the scopes field from extraParameters and transform them into an equivalent expression for security rules`

File: src/PrestaShopBundle/ApiPlatform/Resources/Hook.php
Comment (jolelievre) [inline src/PrestaShopBundle/ApiPlatform/Resources/Hook.php]:
```suggestion
                'scopes' => ['hook_read'],
```

File: app/config/security_test.yml
Comment (jolelievre) [inline app/config/security_test.yml]:
```suggestion
          my_client_id: { password: 'prestashop', roles: [ 'ROLE_HOOK_READ' ] }
```

File: src/PrestaShopBundle/ApiPlatform/Resources/Hook.php
Comment (jolelievre) [inline src/PrestaShopBundle/ApiPlatform/Resources/Hook.php]:
```suggestion
                'scopes' => ['hook_read'],
```

File: src/PrestaShopBundle/ApiPlatform/Resources/Hook.php
Comment (jolelievre) [inline src/PrestaShopBundle/ApiPlatform/Resources/Hook.php]:
```suggestion
                        'description' => 'hook_read <br> hook_write ',
```

File: src/PrestaShopBundle/ApiPlatform/Resources/Hook.php
Comment (jolelievre) [inline src/PrestaShopBundle/ApiPlatform/Resources/Hook.php]:
```suggestion
                    'hook_status_read',
```

File: src/PrestaShopBundle/ApiPlatform/Resources/Hook.php
Comment (jolelievre) [reply src/PrestaShopBundle/ApiPlatform/Resources/Hook.php]: The scope are snake case no `-` allowed Besides we don't define sub scopes for sub parts of the entity for now So even for the status endpoint the scope needed should the same as the one for the whole hook `hook_read`

File: src/PrestaShopBundle/ApiPlatform/Resources/Hook.php
Comment (jolelievre) [reply src/PrestaShopBundle/ApiPlatform/Resources/Hook.php]: For normalization groups though you can use a specific value since it's not a scope, however we should probably use a snake case convention as well and avoid mixing snake case and khebab case

Comment (jolelievre) [reply src/PrestaShopBundle/ApiPlatform/Resources/Hook.php]: Also, I have one question regarding this normalization group, is it really needed? I believe when we discussed this with MAxime he explained to me that these groups were overkill because APIPLatform is able to return a partial DTO when some of its fields are null

The groups were added as an example for normalization but I actually think that if they're not needed we should remove them and keep the endpoints as clean and small and possible. Besides it would validate that we don't need normalization groups to handle partial APIs simply which is actually a more interesting feature than this advanced one based on groups

File: src/PrestaShopBundle/ApiPlatform/Resources/Hook.php
Comment (jolelievre) [inline src/PrestaShopBundle/ApiPlatform/Resources/Hook.php]:
```suggestion
    #[Groups(['hook_status_read'])]
```

File: src/PrestaShopBundle/ApiPlatform/Resources/Hook.php
Comment (jolelievre) [inline src/PrestaShopBundle/ApiPlatform/Resources/Hook.php]:
```suggestion
    #[Groups(['hook_status_read'])]
```

Comment (jolelievre) [reply src/PrestaShopBundle/ApiPlatform/Resources/Hook.php]: Unless we can completely remove those (as explained in previous comment) which would be even better

File: src/PrestaShopBundle/ApiPlatform/Resources/Hook.php
Comment (jolelievre) [inline src/PrestaShopBundle/ApiPlatform/Resources/Hook.php:74]: Maybe we should add the `hook_write` scope here while we're at it no?

**Review by jolelievre — APPROVED** [review]
Thanks @tleon

**Review by boherm — DISMISSED** [review]
LGTM, and tested on my side as asked.
All working in the right way!

**Comment by boherm** [comment]
Before merge, @tleon can you check the failed job in your UI tests: https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/6097740635/job/16579414231 ?

**Review by jolelievre — CHANGES_REQUESTED** [review]
@boherm @tleon Indeed as long as these tests are red we can't merge the PR or it will break all the following UI tests
Besides it's the tests related to API and I restarted them a few times so there are probably some stuff to fix


It's in the tests related to Keycloak so my guess is that the token returned by Keycload has no scopes defined (which makes sense since he doesn't know the, it doesn't provide default scopes and we also don't request for scopes in the token request)

**Review by jolelievre — APPROVED** [review]
Thanks @tleon

**Review by matthieu-rolland — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]
OK for UI Tests

(That disables UI Test for external auth server and will enable it in #33946)

(Ping @boubkerbribri)

**Review by boubkerbribri — APPROVED** [review]

**Review by boherm — APPROVED** [review]

**Comment by florine2623** [comment]
As discussed with @tleon and @boherm , this PR has been QA by dev and is approved ^^
Thanks all !

**Comment by mflasquin** [comment]
thanks @tleon !

**Comment by jolelievre** [comment]
Last run to be sure https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/6234867078

**Comment by jolelievre** [comment]
Last run green (except sanity tests but not related to the PR)

Thanks @tleon

**Pull request merged**
